feat: Add marshaller configuration options for Redis cache adapter

momospnr opened the pull request:
### 概要 / Overview

RedisキャッシュアダプタにMarshallerの設定オプションを追加し、データの圧縮とシリアライズ方式を設定可能にしました。

Add marshaller configuration options to the Redis cache adapter, allowing you to configure data compression and serialization methods.

### 使い方 / Usage

`StorageRedisDsnModule`をインストールする際に`marshallingOptions`を指定します：

Specify `marshallingOptions` when installing `StorageRedisDsnModule`:

```php
use BEAR\QueryRepository\StorageRedisDsnModule;

$this->install(
    new StorageRedisDsnModule(
        dsn: 'redis://localhost:6379',
        marshallingOptions: [
            'enabled' => true,
            'type' => 'deflate',      // 'default' or 'deflate'
            'use_igbinary' => true    // requires ext-igbinary
        ]
    )
);
```

**Marshallerの種類 / Marshaller Types:**
- `default`: PHPの標準シリアライズ（`use_igbinary`を有効にするとより効率的） / Standard PHP serialization (more efficient with `use_igbinary` enabled)
- `deflate`: zlibを使ったデータ圧縮（メモリ使用量削減） / Data compression using zlib (reduces memory usage)

**いつ使うか / When to Use:**

Redisのメモリ使用量を削減したい場合は`deflate`を使用してください。CPU使用率とのトレードオフになります。

Use `deflate` when you want to reduce Redis memory usage. This is a trade-off with CPU usage.

### 変更内容 / Changes

- **新規追加 / New Features**
  - `MarshallerType` Enum: 型安全なmarshaller type定義 / Type-safe marshaller type definition
  - `MarshallerOptions` アノテーション / Annotation: Marshallerオプション設定用のqualifier / Qualifier for marshaller options
  - `MarshallerProvider`: 設定に基づいてMarshallerインスタンスを生成 / Creates marshaller instances based on configuration
  - 包括的なテストスイート / Comprehensive test suite: `MarshallerProviderTest`, `StorageRedisDsnModuleMarshallerTest`

- **機能拡張 / Enhancements**
  - `StorageRedisDsnModule`: `marshallingOptions`, `defaultLifetime`パラメータを追加 / Added parameters
  - RedisTagAwareAdapterにmarshallerとdefaultLifetimeを設定可能 / Configurable marshaller and defaultLifetime
  - PHPDocに詳細な使用例を追加 / Added detailed usage examples in PHPDoc

### サポートするMarshallerタイプ / Supported Marshaller Types

- **default**: 標準のシリアライズ（オプションでigbinaryサポート） / Standard serialization (with optional igbinary support)
- **deflate**: 圧縮機能付きMarshaller（zlibが必要） / Marshaller with compression (requires zlib)

### 技術的な詳細 / Technical Details

- array shapeで型定義を明示化（IDEサポート、静的解析向上） / Explicit type definitions with array shapes (IDE support, improved static analysis)
- Enumによる型安全性の向上（Psalm/PHPStanエラーゼロ） / Enhanced type safety with Enum (zero Psalm/PHPStan errors)
- 不正な値は自動的にDEFAULTにフォールバック / Invalid values automatically fallback to DEFAULT
- 環境変数や設定ファイルからの文字列入力に対応 / Supports string input from environment variables and config files

### 後方互換性 / Backward Compatibility

既存のコードには影響なし。新しいパラメータはすべてオプション（デフォルト値設定済み）。

No impact on existing code. All new parameters are optional with default values.

### 関連ドキュメント / Related Documentation

マニュアルへの追加 / Manual documentation: https://github.com/bearsunday/bearsunday.github.io/pull/315

momospnr committed:
feat: Marshaller options and provider implementation

- 新しいMarshallerOptionsアノテーションを追加し、Redisキャッシュアダプターの設定オプションを注入できるようにしました。
- MarshallerProviderクラスを実装し、設定オプションに基づいてインスタンスを作成する機能を追加しました。
- StorageRedisDsnModuleにMarshallerOptionsをバインドし、オプションを構成できるようにしました。
- MarshallerProviderのテストを追加し、さまざまなタイプの取得を検証しました。
- StorageRedisDsnModuleのテストを追加し、モジュールのバインディングとオプションのデフォルト値を確認しました。

momospnr committed:
feat: Redis DSN module configuration updates

StorageRedisDsnModule.php:
- Redis DSN provider binding added for 'redis'.
- Default lifetime and marshaller bindings updated to use annotations.
- Adjusted parameters for RedisTagAwareAdapter constructor to use annotated bindings.

momospnr committed:
feat: Memcached setup and dependency updates

- テストの依存関係にsodium拡張を追加しました。

momospnr committed:
feat: Sodium marshaller key processing and validation improvements

MarshallerProvider.php:
- 追加: processSodiumKeysメソッドを実装し、Sodium暗号化キーの処理と検証を行う。
- 修正: createSodiumMarshallerメソッドで、キーをprocessSodiumKeysメソッドを使用して処理するように変更。

MarshallerProviderTest.php:
- 追加: Sodiumマシュラーのテストケースを追加し、キーの形式や長さの検証を行うテストを実装。

momospnr committed:
feat: Sodium marshaller関連のコードを削除

MarshallerProvider.phpからSodiumMarshallerに関連するコードを削除しました。これにより、Sodiumマシュラーのサポートが無効化され、関連するオプションやメソッドも削除されました。

MarshallerProviderTest.phpからもSodiumMarshallerに関するテストを削除しました。これにより、Sodiumに関連するテストケースが全て削除され、テストの簡素化が図られました。

momospnr committed:
feat: MarshallerProviderのテストケースを追加

- igbinary拡張が利用できない場合のデフォルトマールシャラーのテストを追加
- Deflateタイプとigbinaryの組み合わせに関するテストを追加
- Deflateタイプでigbinaryが利用できない場合のテストを追加
- 非文字列タイプが指定された場合のデフォルトマールシャラーのテストを追加
- タイプが指定されていない場合のデフォルトマールシャラーのテストを追加

momospnr committed:
feat: RedisとMemcachedの拡張機能を更新

- continuous-integration.yml: sodium拡張を削除し、redisおよびmemcached拡張のみを残しました。

momospnr committed:
feat: MarshallerProviderのエラーハンドリングを改善

MarshallerProvider.php:
- igbinaryおよびzlib拡張の存在チェックを削除し、RuntimeExceptionのスローを無効化しました。

MarshallerProviderTest.php:
- zlibおよびigbinary拡張が利用できない場合のテストケースを削除しました。

koriym commented on src/MarshallerProvider.php:
```diff
@@ -0,0 +1,75 @@
+ */
+final class MarshallerProvider implements ProviderInterface
+{
+    /** @param array<string, mixed> $options Marshalling options */
```
`array<string, mixed>`ではなく、$optionsに何が指定できるか、設定可能なキーを明示的に示すarray shapeにしましょうか。

koriym replied:
Enumにしてcommitしました。

momospnr replied:
@koriym 
修正ありがとうございます！
確かに明示的なarray shapeによる型指定が必要でした。

ドキュメントも更新ありがとうございます

koriym committed:
feat: Use array shape for marshaller options type definition

- Replace array<string, mixed> with explicit array shape for better type safety
- Define allowed keys: enabled, type, use_igbinary with their specific types
- Improve IDE autocompletion and static analysis support
- Keep runtime validation with default case in match expression

koriym committed:
refactor: Use Enum for marshaller type instead of string literals

Introduce MarshallerType enum to replace string-based type handling.

Benefits:
- Type safety: IDE autocomplete and compile-time validation
- Eliminates static analysis issues: No suppressions or baselines needed
- Cleaner match expression: All enum cases are exhaustive, no default case
- Runtime safety: Invalid string values fallback to DEFAULT via tryFrom()
- Backwards compatible: Accepts strings from config files/env variables

This resolves Psalm NoValue/NullArgument errors that occurred with the
previous array shape approach, where the type system considered the
default case unreachable but runtime validation required it.

Changes:
- Add MarshallerType enum with DEFAULT and DEFLATE cases
- Update MarshallerProvider to use enum internally with tryFrom()
- Simplify tests: invalid types now fallback gracefully instead of throwing
- Remove non-string type test (type system prevents this at compile time)
- Remove runtime exception throwing and string type guards
- Update array shape to accept string type (converted to enum internally)

koriym committed:
docs: Add marshaller configuration example to StorageRedisDsnModule PHPDoc

Add usage example and explanation for marshallingOptions parameter in
constructor documentation. This helps developers understand how to
configure compression and serialization options when viewed in IDE.

Example shows:
- How to enable compression with deflate marshaller
- How to use igbinary for efficient serialization
- Brief explanation of available marshaller types

koriym commented:
@momospnr 確認お願いします！
